### PR TITLE
Modified #41008 without GetOverlappedResultEx

### LINKED ISSUE #41007: Error handling is too brittle for Windows IpcStreamFactory/IPC PAL
If a Connect Diagnostic Port doesn't explicitly call `DisconnectNamedPipe` before closing the handle, the error handling in the runtime will erroneously treat this as a generic error rather than a hangup.  The if statement currently only marks `ERROR_PIPE_NOT_CONNECTED` as a hangup, but should also treat `ERROR_PIPE_BROKEN` as a hangup.  The current error path also does not reset the connection, so the cache is still there on subsequent loops through `IpcStreamFactory::GetNextAvailableStream` which leaves us waiting on handles that no longer are attached to anything.

The fix is simple and twofold: reset connections on the error path & map `ERROR_PIPE_BROKEN` to the hangup state rather than error state.  The PR is incoming.

CC @tommcdon @sywhang @noahfalk

### LINKED ISSUE #41217: All libraries tests failing on Win7 x86
All tests are failing in "Libraries Test Run release coreclr Windows_NT x86 Debug" legs on Win7:

```
C:\h\w\B60309FC\w\AC1508F7\e>"C:\h\w\B60309FC\p\dotnet.exe" exec --runtimeconfig Invariant.Tests.runtimeconfig.json --depsfile Invariant.Tests.deps.json xunit.console.dll Invariant.Tests.dll -xml testResults.xml -nologo -nocolor -notrait category=IgnoreForCI -notrait category=OuterLoop -notrait category=failing  
Failed to load the dll from [C:\h\w\B60309FC\p\shared\Microsoft.NETCore.App\6.0.0\coreclr.dll], HRESULT: 0x8007007F
Failed to bind to CoreCLR at 'C:\h\w\B60309FC\p\shared\Microsoft.NETCore.App\6.0.0\'
Failed to create CoreCLR, HRESULT: 0x80008088
----- end Sat 08/22/2020 15:54:30.84 ----- exit code -2147450743 ----------------------------------------------------------
 Sat 08/22/2020-15:54:30.87
```

Hit in #41199, #41200, ...

### PR DESCRIPTION
**Pull request by josalem**
fixes #41217 

Same changes as #41008 but without using `GetOverlappedResultEx` which is only available in Windows 8 forward.  This modification does not change the behavior of the patch.

Below is the PR description from #41008.

---
fixes #41007

These changes are intended to be backported to 5.0-rc1 once merged into master.

As described in #41007, the runtime was treating `ERROR_BROKEN_PIPE` as a generic error rather than a hangup error.  These changes correct that misbehavior and reset the connection on generic errors in case additional errors have been missed.  Additionally, they simplify the logic of `IpcStream::{Read|Write}` to use `GetOverlappedResultEx` which directly takes a timeout rather than `GetOverlappedResult` followed by `WaitForSingleObject`.

CC @jander-msft @tommcdon

---

## Customer Impact

This currently impacts `dotnet-monitor` on Windows.

If a Connect Diagnostic Port on Windows doesn't call `DisconnectNamedPipe` before closing it will cause runtimes to be unable to connect to another Connect Diagnostic Port at the same address.  For example, `dotnet-monitor` currently does not call `DisconnectNamedPipe` before closing, so if a user restarts `dotnet-monitor` they won't see any apps show up if they connected before.

## Testing

I have tested these changes locally with `dotnet-monitor` both under a debugger and without the debugger.  These fix the observed issue.  I did not add an individual test for this scenario, but one could be created.

## Risk

This is a low risk patch.  It simply improves the error handling and corrects a misbehavior.

### CODE REVIEW

File: src/coreclr/src/debug/debug-pal/win/diagnosticsipc.cpp [hunk 1]
```diff
@@ -488,40 +492,43 @@ bool IpcStream::Write(const void *lpBuffer, const uint32_t nBytesToWrite, uint32
...
+                // or timeout
                 DWORD dwWait = WaitForSingleObject(_oOverlap.hEvent, (DWORD)timeoutMs);
                 if (dwWait == WAIT_OBJECT_0)
                 {
-                    // get the result
-                    fSuccess = GetOverlappedResult(_hPipe,
-                                                   overlap,
-                                                   &nNumberOfBytesWritten,
-                                                   true) != 0;
+                    // async IO compelted, get the result
```
Comment (jkotas) [inline src/coreclr/src/debug/debug-pal/win/diagnosticsipc.cpp:513]:
```suggestion
                    // async IO completed, get the result
```

File: src/coreclr/src/debug/debug-pal/win/diagnosticsipc.cpp [hunk 2]
```diff
@@ -434,39 +434,43 @@ bool IpcStream::Read(void *lpBuffer, const uint32_t nBytesToRead, uint32_t &nByt
...
+                // or timeout
                 DWORD dwWait = WaitForSingleObject(_oOverlap.hEvent, (DWORD)timeoutMs);
                 if (dwWait == WAIT_OBJECT_0)
                 {
-                    // get the result
-                    fSuccess = GetOverlappedResult(_hPipe,
-                                                   overlap,
-                                                   &nNumberOfBytesRead,
-                                                   true) != 0;
+                    // async IO compelted, get the result
```
Comment (jkotas) [inline src/coreclr/src/debug/debug-pal/win/diagnosticsipc.cpp:455]:
```suggestion
                    // async IO completed, get the result
```

**Comment by josalem** [comment]
coreclr pri0 Win-x64 failure is #40916 
libraries Win-x86 failure doesn't seem to have a specific issue for it, but AzDO says the System.Text.Json workitem has been failing as far back as August 16th.  This specific failure looks like a 15 minute timeout on the workitem.

These failures don't look related to these changes; I'll rerun these two jobs.

File: src/libraries/System.Private.CoreLib/src/System/WeakReference.cs [hunk 3]
```diff
@@ -9,7 +9,7 @@ namespace System
     [System.Runtime.CompilerServices.TypeForwardedFrom("mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089")]
     public partial class WeakReference : ISerializable
     {
-        // If you fix bugs here, please fix them in WeakReference<T> at the same time.
+        // If you fix bugs here, please fix them in WeakReference<T> at the same time
```
Comment (noahfalk) [inline src/libraries/System.Private.CoreLib/src/System/WeakReference.cs:12]: Nit: random comment change appears unrelated?

**Review by noahfalk — APPROVED** [review]
LGTM, thanks @josalem!

File: src/libraries/System.Private.CoreLib/src/System/WeakReference.cs (hunk 3 above)
Comment (josalem) [reply src/libraries/System.Private.CoreLib/src/System/WeakReference.cs:12]: It is unrelated.  I changed a comment to make CI run the libraries subset of the runtime pipeline which includes the windows 7 leg.  I can remove the change before merging.

**Pull request merged**